core/base/temporary_conversion.hpp

```diff
@@ -0,0 +1,227 @@
+#define GKO_CORE_BASE_TEMPORARY_CONVERSION_HPP_
+
+
+#include <tuple>
+
+
+#include <ginkgo/core/base/lin_op.hpp>
+#include <ginkgo/core/base/utils.hpp>
+
+
+namespace gko {
+namespace detail {
+
+
+/**
+ * A convert_back_deleter is a type of deleter that converts and copies the data
+ * to an internally referenced object before performing the deletion.
+ *
+ * The deleter will use the `convert_to` method to perform the conversion, and
+ * then delete the passed object using the `delete` keyword. This kind of
+ * deleter is useful when temporarily converting an object with the intent of
+ * converting it back once it goes out of scope.
+ *
+ * There is also a specialization for constant objects that does not perform the
+ * conversion, since a constant object couldn't have been changed.
+ *
+ * @tparam CopyType  the type of converted object being deleted
+ * @tparam OrigType  the type of converted object to which the data will be
+ *                   converted back
+ */
+template <typename CopyType, typename OrigType>
+class convert_back_deleter {
+public:
+    using pointer = CopyType *;
+    using original_pointer = OrigType *;
+
+    /**
+     * Creates a new deleter object.
+     *
+     * @param original  the origin object to which the data will be converted
+     *                  back before deletion
+     */
+    convert_back_deleter(original_pointer original) : original_{original} {}
+
+    /**
+     * Deletes the object.
+     *
+     * @param ptr  pointer to the object being deleted
+     */
+    void operator()(pointer ptr) const
+    {
+        ptr->convert_to(original_);
+        delete ptr;
+    }
+
+private:
+    original_pointer original_;
+};
+
+
+// specialization for constant objects, no need to convert back something that
+// cannot change
+template <typename CopyType, typename OrigType>
+class convert_back_deleter<const CopyType, const OrigType> {
+public:
+    using pointer = const CopyType *;
+    using original_pointer = const OrigType *;
+    convert_back_deleter(original_pointer) {}
+
+    void operator()(pointer ptr) const { delete ptr; }
+};
+
+
+template <typename... ConversionCandidates>
+struct conversion_helper {
+    template <typename TargetType, typename MaybeConstLinOp>
+    static std::unique_ptr<TargetType, std::function<void(TargetType *)>>
+    convert(MaybeConstLinOp *obj)
+    {
+        return convert_impl<TargetType, MaybeConstLinOp,
+                            ConversionCandidates...>(obj);
+    }
+
+    template <typename TargetType, typename MaybeConstLinOp,
+              typename FirstCandidate, typename... TrailingCandidates>
+    static std::unique_ptr<TargetType, std::function<void(TargetType *)>>
+    convert_impl(MaybeConstLinOp *obj)
+    {
+        using candidate_type =
+            typename std::conditional<std::is_const<MaybeConstLinOp>::value,
+                                      const FirstCandidate,
+                                      FirstCandidate>::type;
```

Comment:
You need to include `type_traits` for this to work. Additionally, since we use C++14, I think this would be better:
```suggestion
            std::conditional_t<std::is_const<MaybeConstLinOp>::value,
                                      const FirstCandidate,
                                      FirstCandidate>;
```